torch/_torch_docs.py

```diff
@@ -6485,6 +6485,11 @@ def merge_dicts(*dicts):
 
 Returns a tensor of size :attr:`size` filled with :attr:`fill_value`.
 
+.. warning::
+    In PyTorch 1.5 integral `fill_value`s will produce a warning if `out`
```

Comment:
do you think this is too conservative for this change?  I feel like this API isn't used that much and maybe we should speed things up.  What is the fbcode usage?

Reply:
@gchanan Mike and I talked this over a bit. I can see the argument for skipping the warning and going straight to breaking if the usage is low, though personally I'm still not sure I'd do it so close to the 1.5 cut. 

Is skipping the warning the "speed things up" you were thinking of? We were trying to channel your worldview but you know, imperfect vessels 😁

Reply:
There are 199 hits searching torch.full in FBCode, which is low compared to other functions (like max) we've looked at. Many of the uses actually specify the dtype, too, and for those that don't an easy fix is to simply write in the dtype argument. 

Stepping back: let's talk about concrete deprecation options. Seems like there are three deprecation actions we can take: (1) warn, like this PR does, (2) error, like this PR proposes we do in 1.6, (3) change, like this PR proposes we do in 1.7. 

Warning (1) is desirable because it gives people an opportunity to change without their network breaking. Erroring (2) is critical for preventing silent failures. And changing (3) is the entire point. If we refuse to accept the possibility of silently breaking a network than I think we have to do (2) and (3). @bhosmer and I constructed the following scenario: we implement the change, a user gets a LongTensor expecting a FloatTensor, they divide the LongTensor by another LongTensor, this performs integral, not true, division and they get a LongTensor result with different values than the FloatTensor they had previously. This LongTensor is then added to another FloatTensor, producing a FloatTensor (as expected), but with different values than before. This is a silent failure.

If we accept that argument, the only speedup to consider is whether we warn or error in 1.5. I lean warn in 1.5 because (A) we're close to the snap and (B) we haven't formalized our guidance for when we don't need to warn. We also, plan, for example, to warn with torch.div performing integer division in 1.5, cause it to error in 1.6, and change torch.div's behavior in 1.7.

Lastly, we can try to have our cake and eat it, too. That is, we could implement warn as our baseline, then look at a PR that causes the behavior to error and changes the necessary call sites. If that PR comes together in time and we think it's OK to skip a warning release we could land it.

Reply:
@bhosmer @mrubbery: good discussion.

I'm not sure I would call warning -> changed behavior a silent change of behavior, because it's literally not silent :).  But it's obviously not ideal either, I'm just concerned that 6 months is a _long_ time over which to be making these changes.  I can only speak for myself, but I don't use torch.full because I don't want to have to think about broken type inference and 6 months is a long time to wait.

I also buy the argument that we are too close to the release branch cut to be confident in making it an error.

I'd propose we "try to have our cake and eat it, too":
1) land this with just some ambiguity around the timing, i.e. "torch.full is deprecated and it's behavior will change in a later release"
2) depending on priorities, we try to upgrade the warning to an error in 1.5.  Honestly, 200 occurrences is more than I was expecting; it would be nice to have the number that don't specify a dtype.

Sound good?

Reply:
I updated the documentation and warning to be vaguer about future plans.